daml-lf/spec/daml-lf-1.rst

```diff
@@ -867,6 +867,7 @@ to remove dependence on the order of fields ::
    |Mod:S τ₁ … τₙ|   ↠   σ[α₁ ↦ σ₁, …, αₙ ↦ σₙ]
 
    τ₁ ↠ σ₁   ⋯   τₙ  ↠  σₙ
+   f₁, …, fₘ are mutually distinct
```

Comment:
This isn't necessary, since the `TyStruct` rule enforces distinct field names (`f₁ < … < fₙ lexicographically`), and sorting would preserve the non-uniqueness of field names.

Reply:
Doesn't the `TyStruct` rule only speak about _well-formed_ types? By adding this condition we make sure that this rule only applied if the resulting type is well-formed. Thus, it would transform ill-formed structural records into well-formed ones.

Reply:
`TyStruct` and all the other rules in that section are what *defines* well-formed types. When we sort a badly-ordered struct into a well-ordered struct, we may still have duplicate fields. At that point, `TyStruct` will say "hey, that's not well-formed" and reject the struct.

Reply:
This type normalization rules are operating on pure syntax. They make no well-formedness guarantees. E.g. `'Numeric' 'Int64'` is a possible outcome of type normalization, even though it's not well-formed.

Reply:
We don't have a requirement anywhere that an expression is only well-formed if its type is well-formed, do we? Without this requirement or the condition I added here, the expression
```
< a = 1, a = 2>
```
would be well-formed but have an ill-formed type.

My understanding was that the typing rules are such that you can _prove_ that all well-formed expressions have a well-formed type rather than well-formedness of the type being a requirement. Am I mistaken here?

Reply:
Ah, that sounds like a missing hypothesis in `ExpStructCon`. It should require `σ` be well-formed.

Reply:
Compare e.g. with `ExpVarCon` where each ` σ₁  ↠  σ₁'` has an accompanying `Γ  ⊢  σ₁' : k₁`, for example.

Reply:
That's a good point. I'll fix the `ExpStructCon` rule then.